ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -67,10 +69,34 @@ public BlobId(String id, ClusterMap clusterMap)
    */
   public BlobId(DataInputStream stream, ClusterMap clusterMap)
       throws IOException {
+    this(stream, clusterMap, false);
+  }
+
+  /**
+   * Re-constructs existing blobId by deserializing from data input stream. This constructor includes an optional check
+   * that the stream has no more available bytes after reading.
+   *
+   * @param stream from which to deserialize the blobid
+   * @param clusterMap of the cluster that the blob id belongs to
+   * @param ensureFullyRead {@code true} if the stream should have no more available bytes after deserializing the blob
+   *                        ID.
+   * @throws IOException
+   */
+  private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead)
+      throws IOException {
     this.version = stream.readShort();
     if (version == 1) {
-      this.partitionId = clusterMap.getPartitionIdFromStream(stream);
+      partitionId = clusterMap.getPartitionIdFromStream(stream);
+      if (partitionId == null) {
+        throw new IllegalArgumentException("Partition ID cannot be null");
+      }
       uuid = Utils.readIntString(stream);
+      if (uuid == null) {
+        throw new IllegalArgumentException("Could not parse uuid");
+      }
+      if (ensureFullyRead && stream.available() > 0) {
```

Comment:
For `ByteBufferInputStream` it does, but not in all cases. Because we are expecting the stream to be fully read, I can switch to doing a read.